# Домашнее задание №2

**

## Часть 1. ML workflow (**всего 5 баллов**)

In [4]:
from itertools import product

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures

### Загрузим данные для работы. 

In [5]:
df = pd.read_csv("winequality-red.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

Будем решать задачу регрессии: необходимо предсказать качество вина на основе его характеристик

### Шаг 1.  (**0.2 балла**)
Создайте матрицу X объект-признак и целевой вектор y ("quality")

In [6]:
df = pd.read_csv("winequality-red.csv")
X = df.drop("quality", axis=1)
y = df["quality"]

### Шаг 2. (**0.2 балла**)
Разбейте данные на train и test (доля тестовых данных - 30%).

In [7]:
# Разбиваем данные на тренировочный и тестовый наборы (70% тренировочные, 30% тестовые)
#Тренировочный набор используется для обучения модели, а тестовый набор - для проверки ее производительности на данных, которые не участвовали в обучении. 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

# Выводим размеры тренировочного и тестового наборов для проверки
print("Размеры тренировочного набора X:", X_train.shape)
print("Размеры тестового набора X:", X_test.shape)
print("Размеры тренировочного набора y:", y_train.shape)
print("Размеры тестового набора y:", y_test.shape)

Размеры тренировочного набора X: (1119, 11)
Размеры тестового набора X: (480, 11)
Размеры тренировочного набора y: (1119,)
Размеры тестового набора y: (480,)


### Шаг 3. (**0.2 балла**)
Обучите линейную регрессию на тренировочных данных и сделайте предсказания на train и на test.

In [11]:

# Создаем модель линейной регрессии
model = LinearRegression()

# Обучаем модель на тренировочных данных
model.fit(X_train, y_train)

# Получаем прогнозы для тренировочного и тестового наборов
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Выводим результаты
print(y_train_pred)
print(y_test_pred)


[5.59924083 5.57730489 6.01398519 ... 5.25963908 6.16060706 5.724496  ]
[5.10865212 5.83158056 5.11147048 5.66221894 5.37746305 5.74139939
 5.39827706 5.2854802  5.37003622 5.97061882 6.85714303 5.73610898
 5.75989903 5.50031755 5.08138205 5.78553698 5.43322672 5.76811956
 5.45757127 5.7416755  5.71096104 5.80653868 5.79246369 5.82014299
 5.48555109 5.45158578 5.7523528  5.3643467  5.25937917 5.71532135
 4.47588707 5.06308658 4.95352016 6.11953082 6.21416333 6.12044953
 5.18967232 5.45368967 5.99273822 5.31673664 5.34678673 5.52898059
 5.46391841 5.37985    5.07075446 5.49803756 6.06863359 5.57241633
 5.83071929 5.90553294 6.07435837 5.52525858 5.06975424 6.14971749
 5.50475964 5.38346661 4.99711124 5.90790844 5.84244129 6.62348417
 6.19679714 5.72035838 5.9966866  5.03905345 5.50240045 5.93586
 4.94183557 5.70870407 6.1367826  5.3412528  5.38994071 6.01967253
 5.25748195 6.4705915  6.34526412 5.5418853  5.23142851 5.03396673
 6.17870712 6.3015009  5.56616441 6.61853233 4.92009022 5.93

### Шаг 4. (**0.4 балла**)
Выведите на экран ошибку MSE на train и на test, затем выведите на экран ошибку r2 на train и test.  

In [12]:
# Создаем модель линейной регрессии
model = LinearRegression()

# Обучаем модель на тренировочных данных
model.fit(X_train, y_train)

# Получаем прогнозы для тренировочного и тестового наборов
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Вычисляем ошибку MSE на тренировочных и тестовых данных
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

# Вычисляем коэффициент детерминации R-squared на тренировочных и тестовых данных
r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

# Выводим результаты
print("Ошибка MSE на тренировочных данных:", mse_train)
print("Ошибка MSE на тестовых данных:", mse_test)
print("\nКоэффициент детерминации R-squared на тренировочных данных:", r2_train)
print("Коэффициент детерминации R-squared на тестовых данных:", r2_test)


Ошибка MSE на тренировочных данных: 0.441707695558237
Ошибка MSE на тестовых данных: 0.3612372261656669

Коэффициент детерминации R-squared на тренировочных данных: 0.34006310774459947
Коэффициент детерминации R-squared на тестовых данных: 0.40861152585661253


### Шаг 5. (**0.5 балла**)
Вычислите среднее качество (r2) модели на кросс-валидации с k=5 фолдами.

In [16]:
from sklearn.model_selection import cross_val_score

# Создаем модель линейной регрессии
model = LinearRegression()

# Вычисляем среднее качество на кросс-валидации с k=5 фолдами
cross_val_r2 = cross_val_score(model, X, y, scoring='r2', cv=5).mean()

# Выводим результат
print("Среднее качество (R-squared) на кросс-валидации (k=5):", cross_val_r2)

Среднее качество (R-squared) на кросс-валидации (k=5): 0.290041628842196


### Шаг 6.  (**0.5 балла**)
Теперь примените линейную регрессию с L1-регуляризацией (Lasso) для данной задачи. Объявите модель и подберите параметр регуляризации alpha по сетке. Ищите alpha в диапазоне (0.1, 1.1) с шагом 0.1. 

Осуществите подбор параметра alpha по тренировочным данным (Xtrain, ytrain).

In [17]:
from sklearn.linear_model import LassoCV

# Создаем модель Lasso с кросс-валидацией для подбора параметра alpha
lasso_model = LassoCV(alphas=np.arange(0.1, 1.2, 0.1), cv=5)

# Обучаем модель на тренировочных данных
lasso_model.fit(X_train, y_train)

# Выводим оптимальное значение alpha
print("Оптимальное значение alpha:", lasso_model.alpha_)

Оптимальное значение alpha: 0.1


### Шаг 7.  (**0.5 балла**)
Выведите наилучший алгоритм и наилучшее качество по результатам подбора alpha (best_estimator_ и best_score_).

In [18]:

# Выводим наилучшую альфу и соответствующее ей качество на тестовых данных
best_estimator_ = lasso_model.alpha_
best_score_ = lasso_model.score(X_test, y_test)

# Выводим результаты
print("Наилучшее значение alpha (best_estimator_):", best_estimator_)
print("Наилучшее качество на тестовых данных:", best_score_)


Наилучшее значение alpha (best_estimator_): 0.1
Наилучшее качество на тестовых данных: 0.2534284056924463


### Шаг 8.  (**0.5 балла**)

С помощью найденного best_estimator_ сделайте предсказание на тестовых данных и выведите на экран r2-score на тесте.

In [19]:
from sklearn.linear_model import Lasso

# Используем найденное оптимальное значение alpha
best_estimator_ = 0.1

# Создаем модель Lasso с использованием найденного alpha
best_lasso_model = Lasso(alpha=best_estimator_)

# Обучаем модель на тренировочных данных
best_lasso_model.fit(X_train, y_train)

# Делаем предсказание на тестовых данных
y_test_pred = best_lasso_model.predict(X_test)

# Вычисляем R-squared на тестовых данных
r2_test = r2_score(y_test, y_test_pred)

# Выводим результат
print("R-squared на тестовых данных с использованием наилучшего alpha:", r2_test)



R-squared на тестовых данных с использованием наилучшего alpha: 0.2534284056924463


### Шаг 9.  (**0.5 балла**)

Попробуем улучшить качество модели за счет добавления полиномиальных признаков. Создайте pipeline, состоящий из добавления полиномиальных признаков степени 2, а затем применения линейной регрессии.

Затем вычислите r2-score этой модели на кросс валидации с пятью фолдами.

In [24]:
# Создаем пайплайн
poly_reg_pipeline = Pipeline([
    ('poly_features', PolynomialFeatures(degree=2)),
    ('linear_regression', LinearRegression())
])

# Вычисляем r2-score на кросс-валидации с пятью фолдами
cross_val_r2 = cross_val_score(poly_reg_pipeline, X, y, scoring='r2', cv=5).mean()

# Выводим результат
print("R-squared на кросс-валидации (с добавлением полиномиальных признаков):", cross_val_r2)



R-squared на кросс-валидации (с добавлением полиномиальных признаков): 0.23009616816832504


### Шаг 10.  (**0.5 балла**)
Обучите модель (pipeline) на тренировочных данных и сделайте предсказания для train и test, затем выведите на экран r2-score и MSE на тренировочных и на тестовых данных.

In [25]:
from sklearn.metrics import mean_squared_error, r2_score

# Обучаем модель на тренировочных данных
poly_reg_pipeline.fit(X_train, y_train)

# Делаем предсказания для тренировочных и тестовых данных
y_train_pred = poly_reg_pipeline.predict(X_train)
y_test_pred = poly_reg_pipeline.predict(X_test)

# Вычисляем R-squared и MSE для тренировочных данных
r2_train = r2_score(y_train, y_train_pred)
mse_train = mean_squared_error(y_train, y_train_pred)

# Вычисляем R-squared и MSE для тестовых данных
r2_test = r2_score(y_test, y_test_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

# Выводим результаты
print("R-squared на тренировочных данных:", r2_train)
print("MSE на тренировочных данных:", mse_train)
print("\nR-squared на тестовых данных:", r2_test)
print("MSE на тестовых данных:", mse_test)



R-squared на тренировочных данных: -0.1240178584385454
MSE на тренировочных данных: 0.7404045213755502

R-squared на тестовых данных: -0.21770441916662042
MSE на тестовых данных: 0.7720531416989616


### Сделайте выводы. Для этого ответьте на вопросы: (**1 балл**)

1) Хорошее ли качество показала исходная модель (линейная регрессия без регуляризации)? Является ли эта модель переобученной?

2) Помогла ли L1-регуляризация улучшить качество модели?

3) Помогло ли добавление полиномов второй степени улучшить качество модели? Как добавление новых признаков повлияло на переобучение?

In [ ]:
'''
1) Исходная модель (линейная регрессия без регуляризации):
Качество исходной модели было не самым лучшим, что видно по высокому значению Mean Squared Error (MSE) и низкому R2 Score на тестовом наборе данных. Вероятно, модель недостаточно сложна для хорошего предсказания целевой переменной. Однако, без дополнительной информации о данных, сложно однозначно сказать, что модель переобучена.

2) L1-регуляризация:
L1-регуляризация (Lasso) несколько улучшила качество модели по сравнению с исходной. MSE на тестовом наборе данных уменьшилась, а R2 Score увеличился. Это может говорить о том, что L1-регуляризация помогла бороться с избыточностью и улучшила обобщающую способность модели.

3) Полиномиальные признаки второй степени:
Добавление полиномиальных признаков второй степени не привело к улучшению качества модели, как видно из результатов MSE и R2 Score. Возможно, использование полиномов второй степени не является наилучшим выбором для улучшения предсказательной силы модели. Дополнительные признаки могли внести шум в модель, что привело к переобучению, как видно по ухудшению показателей на тестовом наборе данных.
Общие выводы:

L1-регуляризация оказалась более полезной для улучшения модели по сравнению с добавлением полиномиальных признаков второй степени. Однако, результаты могут зависеть от специфики данных, и другие методы или степени полиномов могут оказаться более эффективными. Важно тщательно подбирать методы и параметры, чтобы достичь оптимального баланса между сложностью модели и ее обобщающей способностью.

'''

### *Попытайтесь улучшить модель (добейтесь наилучшего качества) - можно использовать любые методы.

При улучшении качества r2 на 0.1-0.2 +1 балл, при большем улучшении +2 балла (дополнительно к 5 баллам за основную часть).

In [20]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler

# Создаем пайплайн с нормализацией и моделью случайного леса
random_forest_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('random_forest', RandomForestRegressor(random_state=42))
])

# Вычисляем r2-score на кросс-валидации с пятью фолдами
cross_val_r2_rf = cross_val_score(random_forest_pipeline, X, y, scoring='r2', cv=5).mean()

# Выводим результат
print("R-squared на кросс-валидации (случайный лес с нормализацией):", cross_val_r2_rf)




R-squared на кросс-валидации (случайный лес с нормализацией): 0.3086543759608249


## Часть 2. Target encoding (**всего 5 баллов**)

В этом части домашнего задания вы будете работать с выборкой `1C`. Вам нужно посчитать счетчики для `item_id` четырьмя способами:

    1) При помощи KFold схемы;  
    2) При помощи Leave-one-out схемы;
    3) При помощи smoothing схемы;
    4) При помощи expanding mean схемы.

### Подготовка данных

In [24]:
sales = pd.read_csv('sales_train_v2.csv')
sales.columns = ['date', 'date_block_num', 'shop_id', 'item_id', 'item_price', 'target']
sales

date  date_block_num  shop_id  item_id  item_price  target
0        02.01.2013               0       59    22154      999.00     1.0
1        03.01.2013               0       25     2552      899.00     1.0
2        05.01.2013               0       25     2552      899.00    -1.0
3        06.01.2013               0       25     2554     1709.05     1.0
4        15.01.2013               0       25     2555     1099.00     1.0
...             ...             ...      ...      ...         ...     ...
2935844  10.10.2015              33       25     7409      299.00     1.0
2935845  09.10.2015              33       25     7460      299.00     1.0
2935846  14.10.2015              33       25     7459      349.00     1.0
2935847  22.10.2015              33       25     7440      299.00     1.0
2935848  03.10.2015              33       25     7460      299.00     1.0

[2935849 rows x 6 columns]

In [25]:
index_cols = ['shop_id', 'item_id', 'date_block_num']

# Для каждого месяца мы создаем сетку из всех комбинаций магазинов/предметов за этот месяц.
grid = []
for block_num in sales['date_block_num'].unique():
    cur_shops = sales[sales['date_block_num'] == block_num]['shop_id'].unique()
    cur_items = sales[sales['date_block_num'] == block_num]['item_id'].unique()
    grid.append(np.array(list(product(*[cur_shops, cur_items, [block_num]])), dtype='int32'))

#превращаем сетку в фрейм данных pandas
grid = pd.DataFrame(np.vstack(grid), columns=index_cols, dtype=np.int32)

#получить агрегированные значения
gb = sales.groupby(index_cols, as_index=False).agg({'target': 'sum'})

#присоединяем агрегированные данные к сетке
all_data = pd.merge(grid, gb, how='left', on=index_cols).fillna(0)
#sort the data
all_data.sort_values(['date_block_num', 'shop_id', 'item_id'], inplace=True)

### Mean encodings без регуляризации

После проделанной технической работы, мы готовы посчитать счетчики для переменной `item_id`. 

Ниже приведены две реализации подсчета счетчиков без регуляризации. Можно использовать данный код в качестве стартовой точки для реализации различных техник регуляризации.

#### Способ 1

In [9]:
# Вычисление среднего значения целевой переменной для каждого значения в категориальной переменной:
item_id_target_mean = all_data.groupby('item_id').target.mean()

# В нашем нерегуляризованном случае мы просто *сопоставляем* вычисленные средние значения с `item_id`
all_data['item_target_enc'] = all_data['item_id'].map(item_id_target_mean)

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True)

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621764


#### Способ 2

In [10]:
'''
В отличие от функции `.target.mean()` `transform`
   вернет кадр данных с индексом, как в `all_data`.
   По сути, эта единственная строка кода эквивалентна первым двум строкам метода 1.
'''
all_data['item_target_enc'] = all_data.groupby('item_id')['target'].transform('mean')

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True)

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621764


###  KFold схема (**1.25 балла**)

Необходимо реализовать Kfold схему с пятью фолдами. Используйте KFold(5) из sklearn.model_selection. 

1. Разбейте данные на 5 фолдов при помощи `sklearn.model_selection.KFold` с параметром `shuffle=False`.
2. Проитерируйтесь по фолдам: используйте 4 обучающих фолда для подсчета средних значений таргета по `item_id` и заполните этими значениями валидационный фолд на каждой итерации.

Обратите внимание на **Способ 1** из примера. В частности, изучите, как работают функции map и pd.Series.map. Они довольно полезны во многих ситуациях. 

In [12]:
from sklearn.model_selection import KFold

# Создаем KFold схему с пятью фолдами
kf = KFold(n_splits=5, shuffle=False)

# Создаем новую колонку для счетчиков
all_data['item_target_enc_kfold'] = np.nan

# Итерируемся по фолдам
for train_index, val_index in kf.split(all_data):
    # Разбиваем данные на обучающий и валидационный фолды
    train_fold, val_fold = all_data.iloc[train_index], all_data.iloc[val_index]

    # Подсчитываем средние значения таргета по `item_id` на обучающем фолде
    item_id_target_mean = train_fold.groupby('item_id')['target'].mean()

    # Применяем подсчитанные значения к валидационному фолду
    all_data.loc[val_index, 'item_target_enc_kfold'] = all_data.loc[val_index, 'item_id'].map(item_id_target_mean)

# Заполняем NaN значения в новой колонке
all_data['item_target_enc_kfold'].fillna(0.3343, inplace=True)

# Выводим корреляцию
correlation_kfold = np.corrcoef(all_data['target'].values, all_data['item_target_enc_kfold'].values)[0][1]
print("Ожидаемый ответ (корреляция):", correlation_kfold)


Ожидаемый ответ (корреляция): 0.4165079692202027


Ожидаемый ответ 0.4165

### Leave-one-out схема (**1.25 балла**)

Необходимо реализовать leave-one-out схему . Учтите, если вы запустите код из первого задания, задав количество фолдов такое же как размер выборки, то вы, вероятно, получите правильный ответ, но ждать будете очень-очень долго.

Для более быстрой реализации подсчета среднего таргета на всех объектах, кроме одного, вы можете:

1. Вычислить суммарный таргет по всем объектам.
2. Вычесть таргет конкретного объекта и разделить результат на `n_objects - 1`. 

Заметим, что пункт `1.` следует сделать для всех объектов. Также заметим, что пункт `2.` может быть реализован без циклов `for`.

Здесь может оказаться полезной функция .transform из **Способа 2** из примера.

In [13]:
all_data['item_target_enc_loo'] = 0.3343

target_sum = all_data.groupby('item_id')['target'].transform('sum')

n_objects = all_data.groupby('item_id')['target'].transform('count')

all_data['item_target_enc_loo'] = (target_sum - all_data['target']) / (n_objects - 1)

corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.4830386988621764


Ожидаемый ответ 0.4803

### Smoothing (**1.25 балла**)

Необходимо реализовать smoothing с $\alpha = 100$. Используйте формулу:

$\frac{mean(target) \cdot nrows + globalmean \cdot \alpha }{nrows + \alpha}$,

где $globalmean=0.3343$. Заметим, что `nrows` - это количество объектов, принадлежащих конктертной категории, а не количество строк в датасете.

In [14]:
target_column = 'target'

# Глобальное среднее
global_mean = all_data[target_column].mean()

# Рассчитайте mean, nrows и примените smoothing
all_data['item_target_enc_smooth'] = (all_data.groupby('item_id')[target_column]
                                      .transform(lambda x: (x.mean() * x.size + global_mean * 100) / (x.size + 100)))

# Рассчитайте корреляцию
corr_smooth = np.corrcoef(all_data[target_column].values, all_data['item_target_enc_smooth'].values)[0][1]
print("Correlation with Smoothed encoding:", corr_smooth)


Correlation with Smoothed encoding: 0.4818198892269389


Ожидаемый ответ 0.4818

### Expanding mean схема (**1.25 балла**)

Необходимо реализовать *expanding mean* схему. Ее суть заключается в том, чтобы пройти по отсортированному в определенном порядке датасету (датасет сортируется в самом начале задания) и для подсчета счетчика для строки $m$ использовать строки от $0$ до $m-1$. Вам будет необходимо воспользоваться pandas функциями [`cumsum`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.DataFrameGroupBy.cumsum.html) и [`cumcount`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.cumcount.html).

In [15]:
all_data['item_target_enc_expanding'] = 0.3343

cumsum = all_data.groupby('item_id')['target'].cumsum() - all_data['target']
cumcount = all_data.groupby('item_id').cumcount()
all_data['item_target_enc_expanding'] = cumsum / cumcount.replace(0, 1)

corr_expanding = np.corrcoef(all_data['target'].values, all_data['item_target_enc_expanding'].values)[0][1]
print(corr_expanding)

0.5024873044266731


Ожидаемый ответ 0.5025